src/MongoDB.Driver.Core/Core/Clusters/ClusterDescription.cs

```diff
@@ -38,6 +38,40 @@ internal static ClusterDescription CreateInitial(ClusterId clusterId, ClusterCon
                 ClusterType.Unknown,
                 Enumerable.Empty<ServerDescription>());
         }
+
+        // private static methods
+        private static TimeSpan? CalculateLogicalSessionTimeout(ClusterConnectionMode connectionMode, IEnumerable<ServerDescription> servers)
+        {
+            TimeSpan? logicalSessionTimeout = null;
+
+            foreach (var server in SelectServersThatDetermineWhetherSessionsAreSupported(connectionMode, servers))
+            {
+                if (server.LogicalSessionTimeout == null)
+                {
+                    return null;
+                }
+
+                if (logicalSessionTimeout == null || server.LogicalSessionTimeout.Value < logicalSessionTimeout.Value)
+                {
+                    logicalSessionTimeout = server.LogicalSessionTimeout;
+                }
+            }
+
+            return logicalSessionTimeout;
+        }
+
+        private static IEnumerable<ServerDescription> SelectServersThatDetermineWhetherSessionsAreSupported(ClusterConnectionMode connectionMode, IEnumerable<ServerDescription> servers)
```

Comment:
This exact same method is also present in MongClient in High.

Since we can't make public API changes right now we need a copy of the method in both places.